presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.cost;
+
+import io.prestosql.Session;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.planner.TypeProvider;
+import io.prestosql.sql.planner.iterative.Lookup;
+import io.prestosql.sql.planner.plan.TopNNode;
+
+import java.util.Optional;
+
+import static io.prestosql.sql.planner.plan.Patterns.topN;
+
+public class TopNStatsRule
+        extends SimpleStatsRule<TopNNode>
+{
+    private static final Pattern<TopNNode> PATTERN = topN();
+
+    public TopNStatsRule(StatsNormalizer normalizer)
+    {
+        super(normalizer);
+    }
+
+    @Override
+    public Pattern<TopNNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    protected Optional<PlanNodeStatsEstimate> doCalculate(TopNNode node, StatsProvider statsProvider, Lookup lookup, Session session, TypeProvider types)
+    {
+        PlanNodeStatsEstimate sourceStats = statsProvider.getStats(node.getSource());
```

Comment:
I think you should reset nulls fraction to 0 here. Or is it done anyway somewhere else. Can you please add test where input has non-zero null fraction.

Reply:
not sure I follow.
I completely disregard the null fraction as I validate only against the outputRows. adding it to a test won't really affect the changes.
I can change one of the stats in the test to hold fraction other than 0..
btw, I just followed LimitStatsRule.java

Reply:
Reseting nulls fraction to 0 is an overstatement. But IMO nulls handling could be improved here.

Let use two example queries:
```
presto:default> select * from (values 1, NULL, 2, NULL, 3, NULL, 4, NULL, 5) t(a) order by a limit 6;
  a
------
    1
    2
    3
    4
    5
 NULL
```

```
presto:default> select * from (values 1, NULL, 2, NULL, 3, NULL, 4, NULL, 5) t(a) order by a desc limit 6;
  a
------
    5
    4
    3
    2
    1
 NULL
(6 rows)
```

Here we have a `TopN` operator on column `a`. No matter what sort order we are using, `NULLs` will always end up last. So estimated nulls-fraction for an output should be adjusted. Something like this should do:
```
if limit > input.rowsCount {
  return input stats
}

inputNonNullsCount = input.rowCount - input.rowCount * input.nullsFraction;;

outputRowCount = limit
if limit < inputNonNullsCount {
   outputNullsCount = 0
} else {
   outputNullsCount = limit - inputNonNullsCount
}

outputNullsFraction = outputNullsCount/outputRowCount
```

Note that if ordering uses multiple columns, above logic is valid only for the first one.

Reply:
Got you now.
Still now sure how to proceed.
Mainly because I didn't fully understood the first column comment, how do I differ? what the single null fraction value stands for in case of multiple columns?
Regardless,  because one can also use `order by 1 NULLS first` I must look at the sortOrder to re-calculate the null fraction correctly

Reply:
> Mainly because I didn't fully understood the first column comment, how do I differ? what the single null fraction value stands for in case of multiple columns?

There is no single nulls-fraction value. The nulls-fraction is stored per column. You can modify it either using `PlanNodeStatsEstimate.mapSymbolColumnStatistics` or explicitly setting via `PlanNodeStatsEstimate.Builder#addSymbolStatistics()`.
If there are multiple columns used for ordering, I think we cannot derive anything reasonable regarding nulls fraction for any column but first. So I think we should leave source null-fraction for other ones.

> Regardless, because one can also use `order by 1 NULLS first` I must look at the sortOrder to re-calculate the null fraction correctly

Good point.

Reply:
ended up a bit ugly:
 ```
Symbol symbol = node.getOrderingScheme().getOrderBy().get(0); // Assuming not empty list
        SymbolStatsEstimate symbolStatsEstimate = sourceStats.getSymbolStatistics().get(symbol);
        SortOrder sortOrder = node.getOrderingScheme().getOrdering(symbol);
        long count = node.getCount();

        if ((symbolStatsEstimate == null) || (sortOrder == null)) {
            return Optional.of(PlanNodeStatsEstimate.buildFrom(sourceStats)
                                                    .setOutputRowCount(count)
                                                    .build());
        }

        double nullCount = rowCount * symbolStatsEstimate.getNullsFraction();
        double outNullFraction;

        if (sortOrder.isNullsFirst()) {
            if (nullCount > count) {
                outNullFraction = 1;
            }
            else {
                outNullFraction = nullCount / count;
            }
        }
        else {
            double nonNullCount = (rowCount - nullCount);
            if (nonNullCount > count) {
                outNullFraction = 0;
            }
            else {
                outNullFraction = (count - nonNullCount) / count;
            }
        }
        SymbolStatsEstimate newSymbolStatsEstimate = SymbolStatsEstimate.buildFrom(symbolStatsEstimate).setNullsFraction(outNullFraction).build();

        // TopN actually limits (or when there was no row count estimated for source)
        return Optional.of(PlanNodeStatsEstimate.buildFrom(sourceStats)
                .setOutputRowCount(count)
                .addSymbolStatistics(symbol, newSymbolStatsEstimate)
                .build());
    }

```
still think it worth it?

Reply:
You can make it (IMO) a bit cleaner with `mapSymbolColumStatistics`. Like this:
```java
    @Override
    protected Optional<PlanNodeStatsEstimate> doCalculate(TopNNode node, StatsProvider statsProvider, Lookup lookup, Session session, TypeProvider types)
    {
        PlanNodeStatsEstimate sourceStats = statsProvider.getStats(node.getSource());
        if (sourceStats.getOutputRowCount() <= node.getCount()) {
            return Optional.of(sourceStats);
        }

        Symbol firstOrderSymbol = node.getOrderingScheme().getOrderBy().get(0); // Assuming not empty list
        SortOrder sortOrder = node.getOrderingScheme().getOrdering(firstOrderSymbol);
        long limitCount = node.getCount();

        PlanNodeStatsEstimate resultStats = PlanNodeStatsEstimate.buildFrom(sourceStats)
                .setOutputRowCount(limitCount)
                .build();

        // augment null fraction estimation for first ORDER BY symbol
        double rowCount = sourceStats.getOutputRowCount();
        resultStats = resultStats.mapSymbolColumnStatistics(firstOrderSymbol, symbolStats -> {
            SymbolStatsEstimate.Builder newStats = SymbolStatsEstimate.buildFrom(symbolStats);
            double nullCount = rowCount * symbolStats.getNullsFraction();
            if (sortOrder.isNullsFirst()) {
                if (nullCount > limitCount) {
                    newStats.setNullsFraction(1.0);
                }
                else {
                    newStats.setNullsFraction(nullCount / limitCount);
                }
            }
            else {
                double nonNullCount = (rowCount - nullCount);
                if (nonNullCount > limitCount) {
                    newStats.setNullsFraction(0.0);
                }
                else {
                    newStats.setNullsFraction((limitCount - nonNullCount) / limitCount);
                }
            }
            return newStats.build();
        });
        return Optional.of(resultStats);
    }
```

I think it is nice to have. What needs to be added to above is handling division by 0.0. And testing.

Reply:
fixed and pushed.
thanks for the very responsive review.